File: src/main/java/net/flyingfishflash/loremlist/api/data/response/LrmItemResponse.java
package net.flyingfishflash.loremlist.api.data.response;

import java.util.Set;
import java.util.UUID;
import kotlinx.datetime.Instant;
import net.flyingfishflash.loremlist.domain.lrmitem.LrmItem;
import net.flyingfishflash.loremlist.domain.lrmlist.LrmListSuccinct;

/** Representation of a list item outside the context of a list association */
public record LrmItemResponse(
    UUID id,
    String name,
    String description,
    int quantity,
    String owner,
    Instant created,
    String creator,
    Instant updated,
    String updater,
    Set<LrmListSuccinct> lists) {

  public static LrmItemResponse fromLrmItem(LrmItem lrmItem) {
    return new LrmItemResponse(
        lrmItem.id(),
        lrmItem.name(),
        lrmItem.description(),
        0,
        lrmItem.owner(),
        lrmItem.created(),
        lrmItem.creator(),
        lrmItem.updated(),
        lrmItem.updater(),
        lrmItem.lists());
  }
}

File: src/main/kotlin/net/flyingfishflash/loremlist/api/data/response/LrmItemResponse.kt
package net.flyingfishflash.loremlist.api.data.response

import kotlinx.datetime.Instant
import kotlinx.serialization.EncodeDefault
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.Serializable
import net.flyingfishflash.loremlist.core.serialization.UuidSerializer
import net.flyingfishflash.loremlist.domain.lrmitem.LrmItem
import net.flyingfishflash.loremlist.domain.lrmlist.LrmListSuccinct
import java.util.UUID

/**
 * Representation of a list item outside the context of a list association
 */
@Serializable
@OptIn(ExperimentalSerializationApi::class)
data class LrmItemResponse(
  @Serializable(with = UuidSerializer::class)
  val id: UUID,
  val name: String,
  val description: String? = null,
  @EncodeDefault val quantity: Int = 0,
  val owner: String,
  val created: Instant,
  val creator: String,
  val updated: Instant,
  val updater: String,
  val lists: Set<LrmListSuccinct>,
) {
  companion object {
    fun fromLrmItem(lrmItem: LrmItem): LrmItemResponse {
      return LrmItemResponse(
        id = lrmItem.id,
        name = lrmItem.name,
        description = lrmItem.description,
        owner = lrmItem.owner,
        created = lrmItem.created,
        creator = lrmItem.creator,
        updated = lrmItem.updated,
        updater = lrmItem.updater,
        lists = lrmItem.lists,
      )
    }
  }
}
